toolkit/tools/internal/azureblobstorage/azureblobstorage.go

```diff
@@ -148,3 +192,88 @@ func Create(tenantId string, userName string, password string, storageAccount st
 
 	return nil, errors.New("unknown authentication type")
 }
+
+// CreateFromURL creates an AzureBlobStorage client from a storage account URL
+func CreateFromURL(storageAccountURL string) (abs *AzureBlobStorage, err error) {
+	// Parse the URL to extract storage account information
+	storageAccountName, _, _, parseErr := ParseAzureBlobStorageURL(storageAccountURL)
+	if parseErr != nil {
+		return nil, fmt.Errorf("failed to parse storage account URL: %w", parseErr)
+	}
+
+	abs, err = Create("", "", "", storageAccountName, AzureCLIAccess)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Azure Blob Storage client: %w", err)
+	}
+
+	return abs, nil
+}
+
+// DownloadFileWithRetry downloads a file from an Azure Blob Storage using the Azure SDK for Go with retry logic
+// ctx: The context to use for the download. Use context.Background() if no other context is available.
+// azureBlobStorage: The Azure Blob Storage client.
+// srcUrl: The full Azure Blob Storage URL including container and blob path.
+// dstFile: The local file to save the download to.
+// timeout: The maximum duration for the download operation, use 0 for no timeout.
+// returns: wasCancelled: true if the download was cancelled via the context, false otherwise.
+// returns: err: An error if the download failed (including being cancelled), nil otherwise.
+func DownloadFileWithRetry(
+	ctx context.Context,
+	azureBlobStorage *AzureBlobStorage,
+	srcUrl, dstFile string,
+	timeout time.Duration,
+) (wasCancelled bool, err error) {
+	var closeCtx context.CancelFunc
+
+	if ctx == nil {
+		return false, fmt.Errorf("context is nil")
+	}
+
+	if timeout < 0 {
+		return false, fmt.Errorf("invalid timeout: %s", timeout)
+	}
+
+	if timeout == 0 {
+		ctx, closeCtx = context.WithCancel(ctx)
+	} else {
+		ctx, closeCtx = context.WithTimeout(ctx, timeout)
+	}
+	defer closeCtx()
+
+	// Parse the URL to get container and blob names
+	_, containerName, blobName, parseErr := ParseAzureBlobStorageURL(srcUrl)
+	if parseErr != nil {
+		return false, fmt.Errorf("failed to parse source URL: %w", parseErr)
+	}
+
+	logger.Log.Infof("Attempting Azure SDK download for blob: %s/%s", containerName, blobName)
+
+	retryNum := 1
+	errorWas404 := false
+	wasCancelled, err = retry.RunWithDefaultDownloadBackoff(ctx, func() error {
+		netErr := azureBlobStorage.Download(ctx, containerName, blobName, dstFile)
+		if netErr != nil {
+			// Check if the error is a 404-like condition (blob or container not found)
+			if errors.Is(netErr, ErrDownloadInvalidResponse404) {
```

Comment:
When do we set an error to `ErrDownloadInvalidResponse404` or wrap it in another error?